<?xml version="1.0" encoding="utf-8"?>
<?xml version="1.0" encoding="utf-8"?>
<?xml version="1.0" encoding="utf-8"?>
<?xml version="1.0" encoding="utf-8"?>
<?xml version="1.0" encoding="utf-8"?>
<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="control_mode" default="0" />
  <arg name="type" default="quad" />
  <arg name="estimate_mode"  default= "0" />
  <arg name="headless" default="True" />
  <arg name="worldtype" default="$(find ceil_effect_control)/urdf/world/ceiling.world" />
  <arg name="launch_gazebo" default="True" />
  <arg name="spawn_x" default="0.0"/>
  <arg name="spawn_y" default="0.0"/>
  <arg name="spawn_z" default="0.0"/>
  <arg name="spawn_yaw" default="0.0"/>

  ###########  UAV Config  ###########
  # HYDRUS = 16 # 0x10, hydrus type with  4dof underactuated property
  <param name="/uav_info/uav_model" value= "16" />

  ###########  Motor Config  ###########
  <rosparam file="$(find hydrus)/config/MotorInfo.yaml" command="load" />

  ###########  Servo Config  ###########
  <rosparam file="$(find hydrus)/config/$(arg type)/Servo.yaml" command="load" />

  ###########  Battery Config  ###########
  <rosparam file="$(find hydrus)/config/$(arg type)/Battery.yaml" command="load" />

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" output="screen" >

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimator/estimate_mode" value= "2" if="$(arg simulation)" />
    <param name="estimator/estimate_mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>

    # World Pos Control Mode: 0
    # World Vel Control Mode: 2
    # Local Vel Control Mode: 3
    # Attitude Control Mode: 4
    <param name="navigator/xy_control_mode"  value="$(arg control_mode)"/>

    <param name="simulation" value="$(arg simulation)" />  <!--for sensor plugin-->
    <param name="param_verbose" value="false"/>
    <param name="main_rate" type="double" value="40"/>

    ###########  Sensor Fusion  ###########
    <rosparam file="$(find ceil_effect_control)/config/egomotion_estimation/ceil_effect.yaml" command="load" if="$(arg real_machine)"/>
    <rosparam file="$(find hydrus)/config/$(arg type)/egomotion_estimation/simulation.yaml" command="load" if="$(eval arg('simulation') * (1 - arg('real_machine')))"/>
    <remap from="/joint_states" to="/hydrusx/joint_states"/>
    <remap from="/aerial_robot_control_four_axis" to="/aerial_robot_control_four_axis_temp"/>
    ###########  PID Control  ###########
    <rosparam file="$(find hydrus)/config/$(arg type)/DifferentialFlatnessPidControlConfig.yaml" command="load" />

    ###########  Teleop  ###########
    <rosparam file="$(find hydrus)/config/$(arg type)/TeleopNavigationConfig.yaml" command="load" />
  </node>

  ###########  Robot Model  ###########
  <include file="$(find hydrus)/launch/robot_model.launch" >
    <arg name="type" value="$(arg type)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="need_joint_state" value="false" if ="$(eval arg('simulation') + arg('real_machine') > 0)"/>
    <arg name="description_mode" value="gazebo" if="$(arg simulation)"/>
    <arg name="direct_model" value="true" />
    <arg name="direct_model_name" value="$(find ceil_effect_control)/robots/hydrus/ceiling_inspection.urdf.xacro" if="$(eval (1 - arg('simulation')))"/>
    <arg name="direct_model_name" value="$(find ceil_effect_control)/robots/hydrus/ceiling_inspection.gazebo.xacro" if="$(eval arg('simulation'))"/>
  </include >

  ###########  Transform Control  ###########
  <node pkg="hydrus" type="transform_control_node" name="hydrusx"  output="screen" respawn="false">
    <rosparam file="$(find hydrus)/config/$(arg type)/Hydrus.yaml" command="load" />
    <remap from="joint_states" to="hydrusx/joint_states"/>
    <param name="kinematic_verbose" value="false" />
    <param name="control_verbose" value="false" />
    <param name="debug_verbose" value="false" />
    <param name="verbose" value="false"/>
  </node>

  ###########  Sensors  ###########
  <include file="$(find ceil_effect_control)/launch/includes/sensors.launch.xml" >
    <arg name="real_machine" value="$(arg real_machine)" />
    <arg name="simulation" value="$(arg simulation)" />
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="aerial_robot_model" type="servo_bridge_node" name="servo_bridge"  output="screen"  ns="hydrusx" />

  ########## Ceiling Effect Simulation ##########
  <node pkg="ceil_effect_control" type="hydrusx_ceil_effect_simulation.py" name="hydrusx_ceil_effect_node" />

  ########## Simulation in Gazebo #########
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(eval arg('simulation') * (1 - arg('real_machine')))" >
    <arg name="model" default="hydrusx" />
    <arg name="gui" default="false" if="$(arg headless)" />
    <arg name="headless" default="$(arg headless)" />
    <arg name="launch_gazebo" default="$(arg launch_gazebo)" />
    <arg name="worldtype" value="$(arg worldtype)" />
    <arg name="spawn_x" value="$(arg spawn_x)" />
    <arg name="spawn_y" value="$(arg spawn_y)" />
    <arg name="spawn_z" value="$(arg spawn_z)" />
    <arg name="spawn_yaw" value="$(arg spawn_yaw)" />
  </include>

</launch>
